--- src/desktop/frames.c	2027-01-01 00:00:00.000000000 +0000
+++ src/desktop/frames.c	2027-01-01 00:00:00.000000000 +0000
@@ -821,8 +821,29 @@
         else
             sibling_d->value = (sibling_d->value * sibling_dimension) / sibling_pixels;

-        /* todo: the availble resize may have changed, update the drag
-         * box */
+        if (bw->drag.type == DRAGGING_FRAME) {
+            struct rect drag_rect = { 0, 0, 0, 0 };
+
+            if (width) {
+                int min_x = bw->x;
+                int max_x = sibling->x + sibling->width;
+                if (x < min_x) {
+                    bw->drag.start_x = min_x;
+                } else if (x > max_x) {
+                    bw->drag.start_x = max_x;
+                }
+            } else {
+                int min_y = bw->y;
+                int max_y = sibling->y + sibling->height;
+                if (y < min_y) {
+                    bw->drag.start_y = min_y;
+                } else if (y > max_y) {
+                    bw->drag.start_y = max_y;
+                }
+            }
+
+            browser_window_set_drag_type(bw, DRAGGING_FRAME, &drag_rect);
+        }
         return true;
     } else if (sibling_d->unit == FRAME_DIMENSION_RELATIVE) {
         if ((bw_pixels == 0) && (sibling_dimension == 0))
@@ -834,8 +855,29 @@
         else
             bw_d->value = (bw_d->value * bw_dimension) / bw_pixels;

-        /* todo: the availble resize may have changed, update the drag
-         * box */
+        if (bw->drag.type == DRAGGING_FRAME) {
+            struct rect drag_rect = { 0, 0, 0, 0 };
+
+            if (width) {
+                int min_x = bw->x;
+                int max_x = sibling->x + sibling->width;
+                if (x < min_x) {
+                    bw->drag.start_x = min_x;
+                } else if (x > max_x) {
+                    bw->drag.start_x = max_x;
+                }
+            } else {
+                int min_y = bw->y;
+                int max_y = sibling->y + sibling->height;
+                if (y < min_y) {
+                    bw->drag.start_y = min_y;
+                } else if (y > max_y) {
+                    bw->drag.start_y = max_y;
+                }
+            }
+
+            browser_window_set_drag_type(bw, DRAGGING_FRAME, &drag_rect);
+        }
         return true;
     }
