--- frontends/gtk/download.c
+++ frontends/gtk/download.c
@@ -453,6 +453,7 @@
         g_signal_connect_swapped(item, "activate", G_CALLBACK(gtk_button_clicked), dl_ctx.clear);
         gtk_menu_shell_append(GTK_MENU_SHELL(menu), item);

+        g_signal_connect(menu, "deactivate", G_CALLBACK(gtk_widget_destroy), NULL);
         gtk_widget_show_all(GTK_WIDGET(menu));
         nsgtk_menu_popup_at_pointer(menu, NULL);
     }
